src/main/java/org/folio/processing/matching/reader/MarcValueReaderImpl.java

```diff
@@ -0,0 +1,108 @@
+
+/**
+ * Implementation of MatchValueReader for MARC records
+ */
+public class MarcValueReaderImpl implements MatchValueReader {
+
+  private static final String MARC_FIELDS_POINTER = "/fields";
+  private static final String MARC_SUBFIELDS_POINTER = "/subfields";
+  private static final String MARC_IND_1_FIELD_NAME = "ind1";
+  private static final String MARC_IND_2_FIELD_NAME = "ind2";
+  private static final String FIELD_PROFILE_LABEL = "field";
+  private static final String IND_1_PROFILE_LABEL = "indicator1";
+  private static final String IND_2_PROFILE_LABEL = "indicator2";
+  private static final String SUBFIELD_PROFILE_LABEL = "recordSubfield";
+
+  @Override
+  public Value read(EventContext context, MatchDetail matchDetail) {
+    MatchExpression matchExpression = matchDetail.getIncomingMatchExpression();
+    if (matchExpression.getDataValueType() == VALUE_FROM_RECORD) {
+      String marcRecord = context.getObjects().get(MARC.value());
+      return readValueFromRecord(marcRecord, matchExpression);
+    }
+    return MissingValue.getInstance();
+  }
+
+  @Override
+  public boolean isEligibleForEntityType(MatchProfile.IncomingRecordType incomingRecordType) {
+    return incomingRecordType == MARC;
+  }
+
+  private Value readValueFromRecord(String marcRecord, MatchExpression matchExpression) {
+    if (StringUtils.isBlank(marcRecord)) {
+      return MissingValue.getInstance();
+    }
+
+    Map<String, String> matchExpressionFields = getMatchExpressionFields(matchExpression.getFields());
+    List<String> marcFieldValues = new ArrayList<>();
+    readMarcFieldValues(marcRecord, matchExpressionFields)
+      .forEach(marcField -> marcFieldValues.add(
+        marcField.isTextual() ? marcField.textValue() : readSubfieldValue(marcField, matchExpressionFields)));
+    marcFieldValues.remove(null);
+
+    if (marcFieldValues.isEmpty()) {
+      return MissingValue.getInstance();
+    } else if (marcFieldValues.size() == 1) {
+      return StringValue.of(marcFieldValues.get(0));
+    } else {
+      return ListValue.of(marcFieldValues);
+    }
+  }
+
+  private Map<String, String> getMatchExpressionFields(List<Field> fields) {
+    Map<String, String> resultMap = new HashMap<>();
+    fields.forEach(field -> resultMap.put(field.getLabel(), field.getValue()));
+    return resultMap;
+  }
+
+  private List<JsonNode> readMarcFieldValues(String marcRecord, Map<String, String> matchExpressionFields) {
+    try {
+      JsonNode fieldsNode = new ObjectMapper().readTree(marcRecord).at(MARC_FIELDS_POINTER);
+      List<JsonNode> fields = fieldsNode.findValues(matchExpressionFields.get(FIELD_PROFILE_LABEL));
+      return fields.stream()
+        .filter(field -> field.isTextual() || isMatchingIdentifiers(field, matchExpressionFields))
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      throw new RuntimeException("Error reading MARC record", e);
```

Comment:
I think it is better to create separate Exception. @KaterynaSenchenko what do you think about it?

Reply:
agree. Plz add a new Exception and we will use it in case of any reading errors